Add show-match-scores command

WillB97 opened the pull request:
Prints ranking. league and match points for every match.
Matches can be filtered to a single team.

Output format is:
```
$ srcomp show-match-scores dummy-comp HRS
                      |        Zone 0        |        Zone 1        |
 Display Name | Arena | TLA |Rank|Game|League| TLA |Rank|Game|League|
   Match 5    |   A   | BGS |  1 | 10 |    8 | HRS |  4 |  1 |    2 |
   Match 10   |   B   | SCC |  2 |  6 |    6 | DSF |  1 |  8 |    8 |
   Match 16   |   A   | QEH |  1 |  7 |    8 | HRS |  4 |  3 |    2 |
...
```

WillB97 committed:
Add show-match-scores command
Prints ranking. league and match points for every match.
Matches can be filtered to a single team.

WillB97 committed:
Make filter condition more clear

WillB97 committed:
Merge column width calculation loops

WillB97 committed:
Use better method for calculating number of corners

WillB97 committed:
Imporve line formatting

WillB97 committed:
Use srcomp types

WillB97 committed:
Remove failing import

WillB97 committed:
Remove unused constants

WillB97 committed:
Add handling for empty corners

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+    corners: Dict[int, MatchCorner]
+
+
+def collect_match_info(comp, match) -> MatchResult:
```
I expect `mypy` wants the arguments to be typed too.

WillB97 replied:
I've added types in e0a900a0aa3a4d5db9791bafe02334bf57082d37 but I have a type error with the degroup command.
```
sr/comp/cli/show_match_scores.py:47: error: Argument 1 to "degroup" has incompatible type "Mapping[RankedPosition, Set[TLA]]"; expected "Mapping[Optional[LeaguePosition], Iterable[TLA]]"  [arg-type]
```

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+
+    # TODO hide arena column w/ single arena?
+
+    num_teams_per_arena = getattr(comp, 'num_teams_per_arena', len(comp.corners))
```
Why is this `getattr`? `num_teams_per_arena` should always be present.

WillB97 replied:
That line was copied from srcomp-http, I've changed it in 07f25920e39c35e81e73ed614482369d3ec9662a

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+        )
+
+    if len(match_results) == 0:
+        print("Not matches found, TLA may be invalid")
```
```suggestion
        print("No matches found, TLA may be invalid")
```
Though it might be better if we could validate the TLA explicitly.

WillB97 replied:
Fixed the typo in 4d06c95714e3be4a7efd2793c44921a385549aac

WillB97 replied:
c682c2e8111c45c054293601026fbdfc04e13f04 now validates TLAs

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+        print("Not matches found, TLA may be invalid")
+        return
+
+    # Calculate "Display Name" and "Arena" column widths
+    display_name_width = 12  # start with width of label
+    arena_name_width = 5  # start with width of label
```
Please can these lengths not be constants? Even just `len("Arena")` would clarify the intent here.

WillB97 replied:
These the constants have been removed now we're using a library for drawing the table. 8bde289db696a2afd6ca97141f3feecf352ec890

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+    match_results: List[MatchResult] = []
+
+    filter_tla = settings.tla
+    skip_filter = not filter_tla
```
```suggestion
    skip_filter = filter_tla is None
```

WillB97 replied:
Fixed in 8afa4be9a494538e31c431f114bbca95b2888a2b

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+
+class MatchCorner(NamedTuple):
+    tla: TLA
+    ranking: Union[LeaguePosition, str]  # to allow "???" to be used for unknown scores
+    game: Union[GamePoints, str]
+    league: Union[int, str]
```
I'd suggest expressing the unknown case as `None` and force the formatting layer to handle it. That keeps the presentation layer separate from the data layer.
Alternatively, would it make sense to introduce another type so that we have a `ScoredMatch` and a `FutureMatch` (or some better naming)? That means that the type system disallows cases like a match having league points but no game points.

WillB97 replied:
I've moved to using `None` in 14e0af3d04629926b57f751aedcc75eb88e8cddf

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+def print_heading(num_corners, name_width, arena_name_width):
+    print_col("".center(name_width + 1 + arena_name_width))
+    for idx in range(num_corners):
+        print_col(f"Zone {idx}".center(22))
```
Please can we remove this magic number? (what does 22 mean here?) Ditto for the 5 in `"TLA".center(5)` below.

WillB97 replied:
These have been removed now we're using a library for drawing the table. 8bde289db696a2afd6ca97141f3feecf352ec890

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,167 @@
+    print_col(match.display_name.center(name_width))
+    print_col(match.arena.center(arena_name_width))
+
+    for corner in match.corners.values():
```
Feels like this is going to get pretty wide even just for a four zone match. Is this the best layout for a CLI tool?

WillB97 replied:
With the dummy comp the table was 120 wide, this increased to 146 when I moved to using `tabulate` so I've made it wrap each match at 2 zones in 52931ce15f79b651ad4ad36c8261ce46b1422713 which gives a width of 100.

At this point the only part that effects the width of the table is the match and arena names. The next version of `tabulate` looks to be adding and option to force wrap certain columns at a set width.

Alternatively if we can come up with a way to merge the game and league columns they are limited by the length of the heading. In fact omitting all the headers reduces the width to 78.

```
+-----------------------+------+------+------+------+--------+------+------+------+------+--------+
|         Match         | Zone | TLA  | Rank | Game | League | Zone | TLA  | Rank | Game | League |
+-----------------------+------+------+------+------+--------+------+------+------+------+--------+
|     Match 0 in A      |    0 |      |      |      |        |    1 | CLY  |    1 |    9 |      8 |
|                       |    2 | TTN  |    2 |    6 |      6 |    3 |      |      |      |        |
|     Match 0 in B      |    0 | GRS  |    1 |    5 |      8 |    1 | QMC  |    2 |    3 |      6 |
|                       |    2 |      |      |      |        |    3 |      |      |      |        |
|     Match 1 in A      |    0 | WYC  |    4 |    3 |      2 |    1 | QMS  |    2 |    7 |      6 |
|                       |    2 | LSS  |    1 |   10 |      8 |    3 | EMM  |    3 |    5 |      4 |
|     Match 1 in B      |    0 | BPV  |    2 |    6 |      6 |    1 | BDF  |    4 |    2 |      2 |
|                       |    2 | NHS  |    1 |    7 |      8 |    3 | MEA  |    3 |    4 |      4 |
| Quarter 1 (#123) in A |    0 | ???  |   ?? |   ?? |     ?? |    1 | ???  |   ?? |   ?? |     ?? |
|                       |    2 | ???  |   ?? |   ?? |     ?? |    3 | ???  |   ?? |   ?? |     ?? |
| Quarter 2 (#124) in A |    0 | ???  |   ?? |   ?? |     ?? |    1 | ???  |   ?? |   ?? |     ?? |
|                       |    2 | ???  |   ?? |   ?? |     ?? |    3 | ???  |   ?? |   ?? |     ?? |
| Quarter 3 (#125) in A |    0 | ???  |   ?? |   ?? |     ?? |    1 | ???  |   ?? |   ?? |     ?? |
|                       |    2 | ???  |   ?? |   ?? |     ?? |    3 | ???  |   ?? |   ?? |     ?? |
| Quarter 4 (#126) in A |    0 | ???  |   ?? |   ?? |     ?? |    1 | ???  |   ?? |   ?? |     ?? |
|                       |    2 | ???  |   ?? |   ?? |     ?? |    3 | ???  |   ?? |   ?? |     ?? |
```

PeterJCLaw replied:
Hrm. I definitely think we shouldn't mix wrapping with putting things next to each other -- so if we're having a line per zone we should always have that. Both in the sense that the output shouldn't mix the two styles, but also that we shouldn't try to guess which style to use. I can maybe see an argument for letting the user choose, however in that case I think we should ignore the width that that ends up being (since that's then the user's problem -- they got what they asked for).

I think for a first pass here we should go with a simple approach which would always work. This probably means roughly the style you've suggested here, but only one zone wide. (I think this would also simplify the code a bit?)

I think I'd also push the arena name back into its own column? Not sure how that would look in the two-arena case though.

PeterJCLaw reviewed (commented):
Thanks for putting this together, this looks like it could be really useful.
I suggest we work through defining a layout which is going to work for at least four-zone matches before making any other changes as that's likely to need the most change.
If possible I'd like the output to fit within a standard terminal -- 80 characters wide, maybe 100 if we really can't make that work. It feels like it ought to be possible to make that work even for arbitrarily many zones.

WillB97 committed:
Correct typo

WillB97 committed:
is None

WillB97 committed:
Remove unneeded getattr

WillB97 committed:
Use Optional types instead of Union

WillB97 committed:
Solve linting and typing errors

WillB97 committed:
Use tabulate for table display

WillB97 committed:
Make the table more compact

The table now wraps corners at 2, displaying them on the following line. With dummy-comp thins produces a table width of 100.

WillB97 committed:
Add option to limit output to a number of recent matches

WillB97 committed:
Add some comments

WillB97 committed:
Linting

WillB97 committed:
Add typing

WillB97 committed:
Validate TLA

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,233 @@
+    )
+
+
+def match_index(matches: List[MatchSlot], match_num: int) -> int:
+    "Returns the index of the first slot that contains the given match number"
+    for idx, slots in enumerate(matches):
+        for match in slots.values():
+            if match.num == match_num:
+                return idx
+
+    # if no match is found use the last one
+    return idx
```
This shouldn't be necessary -- it should be as simple as `return num` since `matches[num]` should contain matches with that number. If that's not holding then I think we have serious issues elsewhere in SRComp.

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,233 @@
+    displayed_heading = ["Match"]
+
+    for _ in range(num_corners):
+        displayed_heading.append("Zone")
+        displayed_heading.append("TLA")
+        displayed_heading.append("Rank")
+        displayed_heading.append("Game")
+        displayed_heading.append("League")
```
This might be simpler/clearer using `.extend` or `+=`? (The same applies within `generate_displayed_match`)
```suggestion
        displayed_heading += ("Zone", "TLA", "Rank", "Game", "League")
```
Having done that we can also observe a further simplification:
``` python
zone_headers = ("Zone", "TLA", "Rank", "Game", "League")
return ["Match"] + zone_headers * num_corners
```

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,233 @@
+        displayed_corner.append(str(zone))
+        if tla is not None:
+            displayed_corner.append(tla)
+            displayed_corner.append("??" if ranking is None else str(ranking))
+            displayed_corner.append("??" if game is None else str(game))
+            displayed_corner.append("??" if league is None else str(league))
```
I suggest maybe using `-` rather than `??` for unknown score values? It still conveys the unknown nature while being less attention grabbing.

`???` (well, technically we should use the `UNKNOWABLE_TEAM` symbol) should still be used for unknown TLAs though as that's an existing SRComp standard (and is chosen to both visually look like a TLA and indicate the unknowable nature of the value).

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,233 @@
+        skip_filter = False
+        # validate TLA exists
+        if filter_tla not in comp.teams.keys():
+            print('TLA not found')
+            return
```
Since this is an error case, please could we:
- include the "bad" data in the message,
- put the error message on `stderr`, and
- return an error code so that (scripted) callers know something went wrong

I think we have some handlers for this sort of thing in the deploy script if you want some inspiration.
Aside: I've long been meaning to formalise this sort of thing though, likely styled after Django's `CommandError` exception, though that's likely overkill to introduce in this PR.

One approach here might be:
```suggestion
            exit(f"TLA {filter_tla!r} not recognised")
```

PeterJCLaw commented on sr/comp/cli/show_match_scores.py:
```diff
@@ -0,0 +1,233 @@
+        nargs='?',
+        help="filter to matches containing this TLA (ignores --limit)",
+    )
+    parser.add_argument(
+        '--all',
+        action='store_true',
+        help="show all matches (overrides --limit)",
+    )
+    parser.add_argument(
+        '--limit',
+        default=15,
+        help="how many recently scored matches to show (default: %(default)s)",
+    )
```
I believe `argparse` has a way to express mutually exclusive options; please could we use that here?

PeterJCLaw commented:
As a general thought -- I'd probably have this tool just not output anything for the matches which haven't been scored yet. That's how https://studentrobotics.org/comp/points behaves and I'm assuming roughly what you're aiming this to be equivalent to?
This approach avoids needing to work out what should happen to the knockout matches when filtering by TLA, which is a nice way to avoid some complexity.

PeterJCLaw commented:
For when we come back to this -- https://github.com/PeterJCLaw/srcomp/pull/13 likely makes some of the score lookup which is needed here somewhat simpler.
